### PR TITLE
LLM_CHUNK implementation

### PR DESCRIPTION
**Pull request by charleschile**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #18664 

## What this PR does / why we need it:

As part of our document LLM support, we are introducing the `LLM_CHUNK` function. This function can chunk the content in datalink with 4 chunk strategy available. 

**Usage:** `select llm_chunk("<input datalink>", "fixed_width; <width number>");` or  `select llm_chunk("<input datalink>", "<sentence or paragraph or document>");`

**Return Value:** a JSON-like string representation of an array of chunks with offset and size: `[[offset0, size0, "chunk"], [offset1, size1, "chunk"],...]`

Example SQL for fixed with:

```sql
 select llm_chunk(cast('file:///Users/charles/Desktop/codes/testData/example.txt' as datalink), "fixed_width; 11");
```

Example return:

```shell
[[0, 11, "hello world"], [11, 11, " this is a "], [22, 11, "test? hello"], [33, 11, " world! thi"], [44, 11, "s is a test"], [55, 11, ". hello wor"], [66, 3, "ld."]] 
```

Example SQL for sentence:

```sql
 select llm_chunk(cast('file:///Users/charles/Desktop/codes/testData/example.txt' as datalink), "sentence");
```

Example return:

```shell
[[0, 27, "hello world this is a test?"], [27, 13, " hello world!"], [40, 16, " this is a test."], [56, 13, " hello world."]]
```

### CODE REVIEW

**Review by m-schen — APPROVED** [review]

File: pkg/sql/plan/function/func_llm.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		if end > len(runes) {
+			end = len(runes)
+		}
+		chunks = append(chunks, []interface{}{i, end - i, string(runes[i:end])})
+	}
+	return chunks
+}
+
+// Sentence-based chunking
+func sentenceChunk(text string) [][]interface{} {
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:31]: create a struct instead of interface{}.

File: pkg/sql/plan/function/func_llm.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		r, err := ReadFromFile(moUrl, fs)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
+		ctx, err := io.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		if len(ctx) > 65536 /*blob size*/ {
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:127]: how big is 65536?  too small? or just remove the size check.

Comment (charleschile) [reply pkg/sql/plan/function/func_llm.go]: Solved.

**Review by fengttt — CHANGES_REQUESTED** [review]
Please add result verification in tests.

**Review by fengttt — COMMENTED** [review]
How about we add a "trivial" chunking method that will simply chunk the whole doc as 1 chunk?

**Comment by charleschile** [comment]
> How about we add a "trivial" chunking method that will simply chunk the whole doc as 1 chunk?

A new chunk strategy `document` has been supported, it returns the whoe document

**Comment by charleschile** [comment]
> Please add result verification in tests.

result verification has been added.

**Review by fengttt — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

File: pkg/sql/plan/function/func_llm.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+		}
+
+		var input string
+
+		fs := proc.GetFileService()
+		r, err := ReadFromFile(moUrl, fs)
+		if err != nil {
+			return err
+		}
+		defer r.Close()
```
Comment (cpegeric) [inline pkg/sql/plan/function/func_llm.go:152]: defer r.Close() is buggy inside the for loop. variable r only close when end of the function but not each block.  use func() { defer r.Close() } to make sure r is closed at the end of func() {}...   You may run out of file descriptors when rowCount is large. (see sql/plan/function/func_unary.go:WriteFileDatalink() for details)

**Review by cpegeric — CHANGES_REQUESTED** [review]

**Review by m-schen — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Pull request merged**